src/net/socket.rs

```diff
@@ -0,0 +1,229 @@
+                                        self.event_sender.send(SocketEvent::Packet(packet))
+                                    {
+                                        error!("Error sending packet to caller: {:?}", err);
+                                    }
+                                }
+                                Ok(None) => continue,
+                                Err(ref err) => match err.kind() {
+                                    NetworkErrorKind::IOError(io_err)
+                                        if io_err.kind() == io::ErrorKind::WouldBlock =>
+                                    {
+                                        break;
+                                    }
+                                    _ => error!("Error receiving from socket: {:?}", err),
+                                },
+                            };
+                        }
+                    }
+                }
+                _ => unreachable!(),
+            }
+        }
+        Ok(())
+    }
+
+    // Serializes and sends a `Packet` on the socket. On success, returns the number of bytes written.
+    fn send_to(&mut self, packet: Packet) -> NetworkResult<usize> {
+        let connection = self
+            .connections
+            .get_or_insert_connection(packet.addr(), &self.config);
+        let mut packet_data =
+            connection.process_outgoing(packet.payload(), packet.delivery_method())?;
+        let mut bytes_sent = 0;
+
+        if let Some(link_conditioner) = &self.link_conditioner {
+            if link_conditioner.should_send() {
+                for payload in packet_data.parts() {
+                    bytes_sent += self.send_packet(&packet.addr(), &payload)?;
+                }
+            }
+        } else {
+            for payload in connection.gather_dropped_packets() {
+                bytes_sent += self.send_packet(&packet.addr(), &payload)?;
+            }
+
+            for payload in packet_data.parts() {
+                bytes_sent += self.send_packet(&packet.addr(), &payload)?;
+            }
+        }
+
+        Ok(bytes_sent)
+    }
+
+    // Receives a single message from the socket. On success, returns the packet containing origin and data.
+    fn recv_from(&mut self) -> NetworkResult<Option<Packet>> {
+        let (recv_len, address) = self.socket.recv_from(&mut self.recv_buffer)?;
+        if recv_len == 0 {
+            return Err(NetworkErrorKind::ReceivedDataToShort)?;
+        }
+
+        let received_payload = &self.recv_buffer[..recv_len];
+        let connection = self
+            .connections
+            .get_or_insert_connection(address, &self.config);
+        connection.process_incoming(received_payload)
+    }
+
+    // Send a single packet over the UDP socket.
+    fn send_packet(&self, addr: &SocketAddr, payload: &[u8]) -> NetworkResult<usize> {
+        let bytes_sent = self
+            .socket
+            .send_to(payload, addr)
```

Comment:
We don't actually poll to see if this is ready to write too so we should understand what the behvaior is here.

From our discord discussion with @jstnlef:

```
jstnlefToday at 11:18 PM
that won't block
but it "may" error with a "WouldBlock"
if the packet you're trying to send is too large
```

From my understanding, this means this will always write and return right away unless the packet is too large. We should understand the error cases here or buffer and request to write.

Reply:
to add to this it seems we do not handle the `WouldBlock` case, this should always be handled for proper portability https://docs.rs/mio/0.6.16/mio/struct.Poll.html#spurious-events

Reply:
Remind me where we ended up on this one?

Reply:
After discussion I think if send_to fails we need to notify the user that the send did not succeed.